asteroid/models/base_models.py

```diff
@@ -208,34 +215,98 @@ def forward(self, wav):
         """
         # Handle 1D, 2D or n-D inputs
-        was_one_d = False
-        if wav.ndim == 1:
-            was_one_d = True
-            wav = wav.unsqueeze(0).unsqueeze(1)
-        if wav.ndim == 2:
-            wav = wav.unsqueeze(1)
+        was_one_d = wav.ndim == 1
+        # Reshape to (batch, n_mix, time)
+        wav = _unsqueeze_to_3d(wav)
+
         # Real forward
-        tf_rep = self.enc_activation(self.encoder(wav))
+        tf_rep = self.encoder(wav)
+        tf_rep = self.postprocess_encoded(tf_rep)
+        tf_rep = self.enc_activation(tf_rep)
+
         est_masks = self.masker(tf_rep)
+        est_masks = self.postprocess_masks(est_masks)
+
         masked_tf_rep = est_masks * tf_rep.unsqueeze(1)
-        out_wavs = torch_utils.pad_x_to_y(self.decoder(masked_tf_rep), wav)
+        masked_tf_rep = self.postprocess_masked(masked_tf_rep)
+
+        decoded = self.decoder(masked_tf_rep)
+        decoded = self.postprocess_decoded(decoded)
+
+        reconstructed = torch_utils.pad_x_to_y(decoded, wav)
         if was_one_d:
-            return out_wavs.squeeze(0)
-        return out_wavs
+            return reconstructed.squeeze(0)
+        else:
+            return reconstructed
+
+    def postprocess_encoded(self, tf_rep):
+        """Hook to perform transformations on the encoded, time-frequency domain
+        representation (output of the encoder) before encoder activation is applied.
+
+        Args:
+            tf_rep (Tensor of shape (batch, freq, time)):
+                Output of the encoder, before encoder activation is applied.
+
+        Return:
+            Transformed `tf_rep`
+        """
+        return tf_rep
+
+    def postprocess_masks(self, masks):
+        """Hook to perform transformations on the masks (output of the masker) before
+        masks are applied.
+
+        Args:
+            masks (Tensor of shape (batch, n_src, freq, time)):
+                Output of the masker
+
+        Return:
+            Transformed `masks`
+        """
+        return masks
+
+    def postprocess_masked(self, masked_tf_rep):
+        """Hook to perform transformations on the masked time-frequency domain
+        representation (result of masking in the time-frequency domain) before decoding.
+
+        Args:
+            masked_tf_rep (Tensor of shape (batch, n_src, freq, time)):
+                Masked time-frequency representation, before decoding.
+
+        Return:
+            Transformed `masked_tf_rep`
+        """
+        return masked_tf_rep
+
+    def postprocess_decoded(self, decoded):
+        """Hook to perform transformations on the decoded, time domain representation
+        (output of the decoder) before original shape reconstruction.
+
+        Args:
+            decoded (Tensor of shape (batch, n_src, time)):
+                Output of the decoder, before original shape reconstruction.
+
+        Return:
+            Transformed `decoded`
+        """
+        return decoded
```

Comment:
I love this ! 
@popcornell do you feel like to give your opinion on this as well?